cloudimg/aws.py

```diff
@@ -643,3 +703,92 @@ def tag_image(self, image, tags):
         res = image.create_tags(**attrs)
         log.debug("Tag image \"%s\" results: %s" % image.name, res)
         return res
+
+    def deregister_image(self, image):
+        """
+        Deregisters AMI image from AWS.
+
+        Args:
+            image (Image): An EC2 Image
+
+        Returns:
+            Id of deregistered image. (str)
+        """
+        image_id, image_name = image.id, image.name
+        logging.info("Deregistering image %s (%s)", image_id, image_name)
+
+        res = image.deregister()
+
+        logging.debug("Deregister image %s (%s) results: %s",
+                      image_id, image_name, res)
+
+        return image_id
+
+    def delete_snapshot(self, snapshot):
+        """
+        Deletes snapshot from AWS.
+
+        Args:
+            snapshot (Snapshot): An EC2 Snapshot
+
+        Returns:
+            Id of deleted snapshot. (str)
+        """
+        snapshot_id = snapshot.id
+        logging.info("Deleting snapshot %s", snapshot_id)
+
+        res = snapshot.delete()
+
+        logging.debug("Deleted snapshot %s results: %s", snapshot_id, res)
+
+        return snapshot_id
+
+    def delete(self, metadata):
+        """
+        Deletes AMI images and snapshot for given metadata.
+
+        If specified image is not found, we try to find snapshot
+        by other provided metdadata.
+
+        Args:
+            metadata (AWSDeleteMetadata): Metadata about the image
+
+        Returns:
+            tuple (image_id[Str], snapshot_id[Str]]) of removed objects.
+        """
+        deleted_image_id = None
+        deleted_snapshot_id = None
+
+        log.info('Searching for image: %s', metadata.image_id)
+        image = (
+            self.get_image_by_id(metadata.image_id) or
+            self.get_image_by_name(metadata.image_name)
+        )
+
+        if image:
+            # extract snapshot_id from existing image
+            snapshot_id = image.block_device_mappings[0]["Ebs"]["SnapshotId"]
```

Comment:
I'd be a bit cautious here as the image is not always guaranteed to have a snapshot. Maybe catch `KeyError`, `IndexError` here or check if `image.block_device_mappings` is not empty.